src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of size " + size);
+        } else {
+            if (index == size) {
+                add(value);
+                return;
+            }
+            Node<T> next = takeElementByIndex(index);
+            if (next == null) {
+                add(value);
+                return;
+            }
+            Node<T> prev = next.prev;
+            Node<T> newNode = new Node<>(prev, value, next);
+            next.prev = newNode;
+            if (prev == null) {
+                head = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T values : list) {
+            add(values);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> elementByIndex = takeElementByIndex(index);
+        return (elementByIndex == null) ? null : elementByIndex.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> elementByIndex = takeElementByIndex(index);
```

Comment:
```suggestion
        Node<T> element = takeElementByIndex(index);
```
We have only one element in the method, so, a simple name is good enough.